Add SuperRare trades

catherine-2 opened the pull request:
Hi! This is my first non-static table PR so please let me know if there are things I need to fix / any questions! I looked at other trade table files to try to match the format so we can add SuperRare trades.

Brief comments on the purpose of your changes:


*For Dune Engine V2*
I've checked that:
General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)
* [ ] if adding a new materialized table, I edited the optimize table macro

Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

dune-eng commented:
Workflow run id 3137482797 approved.

dune-eng commented:
Workflow run id 3137531660 approved.

soispoke commented:
Hi @catherine-2, thanks a lot for you submission :) 

So for NFT platforms, we actually have model we want to follow so we can integrate all the relevant info into the meta table nft.trades/events/mints/burns...
It would be awesome if you could take a look at other similar spells like LooksRare (https://github.com/duneanalytics/spellbook/tree/main/models/looksrare/ethereum) or others if possible ?

hildobby commented:
Oh I had this one almost done except royalties (was going to work on those before submitting). I planned on finishing it after Zora and Rarible abstractions are merged, let me know if you want to finish it or you want me to take over @catherine-2 :)

catherine-2 commented:
Thank you @soispoke ! Got it - I followed the NFT trades format but will go through and update to make sure it works with all of the different NFT tables so we can add across each. 

@hildobby I will update and tag you to take a look before we merge if that works! I would prefer to work off of this -- spent a lot of time verifying every trade is showing up in this query across all of the different contracts + added logic for RARE trades.

dune-eng commented:
Workflow run id 3185583184 approved.

dune-eng commented:
Workflow run id 3185676119 approved.

catherine-2 commented:
@soispoke  I think this is ready for re-review! Summary of updates:

- I updated the file structure to match LooksRare and other marketplaces
- Built out the events table which populates fees + trades (and flows into the master nft. tables for each of these)
- Added logic for fees + royalties which weren't previously included
- Updated to use sources instead of the dune.com table names 

I verified that this code is working correctly in a V2 query on dune.com Please let me know if anything looks off. Thank you!

soispoke commented on models/nft/nft_events.sql:
```diff
@@ -369,6 +369,46 @@ FROM
+                royalty_fee_amount_usd,
+                royalty_fee_percentage,
+                unique_trade_id
+        FROM {{ ref('superrare_events') }}
```
replace by `superrare_ethereum_events`

soispoke commented on models/nft/nft_fees.sql:
```diff
@@ -254,6 +254,41 @@ FROM
+                tx_from,
+                tx_to,
+                unique_trade_id
+        FROM {{ ref('superrare_fees') }}
```
replace by `superrare_ethereum_fees`

soispoke commented on models/nft/nft_trades.sql:
```diff
@@ -279,6 +279,36 @@ FROM
+                tx_from,
+                tx_to,
+                unique_trade_id
+        FROM {{ ref('superrare_trades') }}
```
replace by `superrare_ethereum_trades`

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,344 @@
+                    else  token_sold_amount*1.0/nullif(token_bought_amount,0)
+                end as eth_per_rare
+        -- from dex.trades
+        from {{ source('dex','trades') }}
```
replace by `{{ ref('dex_trades') }}`

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,344 @@
+            , evt_tx_hash
+            , `to` 
+    -- from erc721_ethereum.evt_Transfer
+    from {{ source('erc721_ethereum','evt_Transfer') }}
```
replace by `{{ source('erc721_ethereum','evt_transfer') }}`

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,344 @@
+                else '' end as auction_house_flag
+            , ROW_NUMBER() OVER (PARTITION BY contract_address, `tokenId` ORDER BY evt_block_time DESC) AS transaction_rank
+    -- from erc721_ethereum.evt_Transfer
+    from {{ source('erc721_ethereum','evt_Transfer') }}
```
replace by `{{ source('erc721_ethereum','evt_transfer') }}`

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,344 @@
+            , `value` as `tokenId`
+            , `to` as mint_address 
+    -- from erc20_ethereum.evt_Transfer
+    from {{ source('erc20_ethereum','evt_Transfer') }}
```
'erc20_ethereum','evt_transfer'

soispoke commented on models/superrare/superrare_ethereum_schema.yml:
```diff
@@ -0,0 +1,210 @@
+        description: "Symbol of the token in which fees are paid out"
+      - &unique_trade_id
+        name: unique_trade_id
+        description:  "Unique trade ID"
```
add a:        
`tests:
    - unique`

to this column

soispoke reviewed (changes requested):
Hi @catherine-2 ! There are still a few small issues, I made comments on the specific files/models but one thing that's also important to add now is incremental filters on the `superrare_ethereum_events` model 🙏

Also, please check the output of the github actions (dbt slim ci) to make sure everything looks good :)

catherine-2 commented:
@soispoke  Thank you for going through and flagging each of the source / ref names that needed to be updated! Sorry about that. 

Just pushed a change to update all of those + add incremental filters. Let me know if there is anything else I should update. Thank you!

soispoke reviewed (commented):
No problem, thanks @catherine-2 !

You also need to add your project in the dbt_project.yml file (as well as fix the issue I commented about) :)

dune-eng commented:
Workflow run id 3191176783 approved.

catherine-2 commented:
@soispoke  I think those are all resolved now! Sorry for all of the minor edits needed on this -- my first non-static table spell :) will make sure to follow all of these practices in future PRs!

dune-eng commented:
Workflow run id 3191320228 approved.

soispoke commented:
Hi @catherine-2, no problem at all very happy that you're learning and contributing to spellbook !

I wanted to push a couple of small changes that still need to be implemented but it seems like I can't do that because I don"t have access to your branch.
Do you mind taking a look at whether you can either accept the changes (I did commit them) or give me permissions ? Otherwise I'll just write it all in another round of review

catherine-2 commented:
@soispoke Where can I see your commits? I already had the box checked here to allow edits to my branch from others but don't see your changes. Let me know if there is another permission I need to update!

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,372 @@
+                end as eth_per_rare
+        -- from dex.trades
+        from {{ ref('dex_trades') }}
+        where   -- RARE trades
```
`{% if is_incremental() %} -- this filter will only be applied on an incremental run
        where block_time >= date_trunc("day", now() - interval '1 week')
        {% endif %}
        and   -- RARE trades`

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,372 @@
+            , `to` 
+    -- from erc721_ethereum.evt_Transfer
+    from {{ source('erc721_ethereum','evt_transfer') }}
+    where contract_address in (select distinct contract_address from all_superrare_sales)
```
{% if is_incremental() %} -- this filter will only be applied on an incremental run
    and evt_block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,372 @@
+            , hash 
+    -- from ethereum.transactions
+    from {{ source('ethereum','transactions') }}
+    ) b 
```
{% if is_incremental() %} -- this filter will only be applied on an incremental run
    where block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,372 @@
+    -- from erc20_ethereum.evt_Transfer
+    from {{ source('erc20_ethereum','evt_transfer') }}
+    where contract_address in (select distinct contract_address from all_superrare_sales)
+    and `value` in (select distinct `_tokenId` from all_superrare_sales)
```
{% if is_incremental() %} -- this filter will only be applied on an incremental run
    and evt_block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,372 @@
+    from {{ source('erc721_ethereum','evt_transfer') }}
+    where contract_address in (select distinct contract_address from token_sold_from_auction)
+    and `tokenId` in (select distinct `_tokenId` from token_sold_from_auction)
+    order by 1, 2, 8
```
{% if is_incremental() %} -- this filter will only be applied on an incremental run
    and evt_block_time >= date_trunc("day", now() - interval '1 week')
    {% endif %}

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,372 @@
+left outer join (select minute, price 
+                -- from prices.usd 
+                from {{ source('prices','usd') }}
+                where blockchain = 'ethereum' and symbol = 'WETH' ) ep 
```
where blockchain = 'ethereum' and symbol = 'WETH'
                {% if is_incremental() %} -- this filter will only be applied on an incremental run
                and minute >= date_trunc("day", now() - interval '1 week')
                {% endif %}) ep

soispoke commented on models/superrare/superrare_ethereum_events.sql:
```diff
@@ -0,0 +1,372 @@
+-- left outer join ethereum.transactions t 
+left outer join  {{ source('ethereum','transactions') }} t
+on a.evt_tx_hash = t.hash
+
```
{% if is_incremental() %} -- this filter will only be applied on an incremental run
and t.block_time >= date_trunc("day", now() - interval '1 week')
{% endif %}

soispoke commented on dbt_project.yml:
```diff
@@ -114,6 +114,12 @@ models:
+    superrare:
+      +schema: superrare
+      +materialized: view
+      ethereum:
```
delete the ethereum part here (nothing under superrare/ethereum path

soispoke reviewed (commented):
ok I just went ahead and added the changes needed afterall :)

dune-eng commented:
Workflow run id 3198861906 approved.

soispoke commented:
@catherine-2 The dbt slim ci is still failing because of some syntax errors that were introduced by recent commits, Idk if I have been clear enough but by clicking on "details" you can see what errors you need to fix
<img width="860" alt="Capture d’écran 2022-10-06 à 17 35 38" src="https://user-images.githubusercontent.com/66172107/194369512-abef7332-a833-4f84-ba1b-f35371bdc1a9.png">

soispoke commented:
Looking at previous messages I should've said that earlier 😅 does it make sense to you ?

catherine-2 commented:
Yep it's showing this error -- I'm still trying to figure out what it is referring to 😅  will fix 

16:34:10  Runtime Error in model superrare_ethereum_events (models/superrare/superrare_ethereum_events.sql)
16:34:10    [PARSE_SYNTAX_ERROR] Syntax error at or near '='(line 164, pos 38)

dune-eng commented:
Workflow run id 3199016203 approved.

dune-eng commented:
Workflow run id 3199066969 approved.

catherine-2 commented:
@soispoke  success! fixed the error

soispoke commented:
@catherine-2 awesome, thanks ! 🎉 everything lgtm now :)

catherine-2 commented:
🎉 ❤️  thank you @soispoke !

@hildobby FYI this was just merged -- let me know if anything looks off on your end!
